### PR TITLE
#90: fixed bugs in sonar findings

### PR DESCRIPTION
**Pull request by AnastasiiaSergienko**

### CODE REVIEW

File: jdbc-adapter/README.md [hunk 1]
```diff
@@ -2,6 +2,20 @@
...
+[![Technical Debt](https://sonarcloud.io/api/project_badges/measure?project=com.exasol%3Avirtualschema-jdbc-adapter-main&metric=sqale_index)](https://sonarcloud.io/dashboard?id=com.exasol%3Avirtualschema-jdbc-adapter-main)
+
+[![Code Smells](https://sonarcloud.io/api/project_badges/measure?project=com.exasol%3Avirtualschema-jdbc-adapter-main&metric=code_smells)](https://sonarcloud.io/dashboard?id=com.exasol%3Avirtualschema-jdbc-adapter-main)
+[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=com.exasol%3Avirtualschema-jdbc-adapter-main&metric=coverage)](https://sonarcloud.io/dashboard?id=com.exasol%3Avirtualschema-jdbc-adapter-main)
+[![Duplicated Lines (%)](https://sonarcloud.io/api/project_badges/measure?project=com.exasol%3Avirtualschema-jdbc-adapter-main&metric=duplicated_lines_density)](https://sonarcloud.io/dashboard?id=com.exasol%3Avirtualschema-jdbc-adapter-main)
+[![Lines of Code](https://sonarcloud.io/api/project_badges/measure?project=com.exasol%3Avirtualschema-jdbc-adapter-main&metric=ncloc)](https://sonarcloud.io/dashboard?id=com.exasol%3Avirtualschema-jdbc-adapter-main)
```
Comment (redcatbear) [inline jdbc-adapter/README.md]: :+1:

File: jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/dialects/SqlDialects.java [hunk 2]
```diff
@@ -127,15 +127,17 @@ public SqlDialect getDialectInstanceForNameWithContext(final String name, final
 
     private SqlDialect instantiateDialect(final String name, final Optional<Class<? extends SqlDialect>> foundDialect,
             final SqlDialectContext context) throws SqlDialectsRegistryException {
-        SqlDialect instance;
         try {
-            final Class<? extends SqlDialect> dialectClass = foundDialect.get();
-            instance = dialectClass.getConstructor(SqlDialectContext.class).newInstance(context);
+            if (foundDialect.isPresent()) {
```
Comment (redcatbear) [inline jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/dialects/SqlDialects.java]: Move the if outside the try.

File: jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/dialects/SqlDialects.java [hunk 3]
```diff
@@ -127,15 +127,17 @@ public SqlDialect getDialectInstanceForNameWithContext(final String name, final
...
+                final Class<? extends SqlDialect> dialectClass = foundDialect.get();
+                return dialectClass.getConstructor(SqlDialectContext.class).newInstance(context);
+            } else {
+                throw new InstantiationException();
```
Comment (redcatbear) [inline jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/dialects/SqlDialects.java]: Say that no dialect with that name exists and give users a list of supported dialect names.

File: jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/JdbcAdapter.java [hunk 4]
```diff
@@ -342,22 +347,21 @@ private static String getUserAndPasswordForImport(ExaMetadata exaMeta, SchemaMet
...
-            final String pushdownQuery, final SqlDialect dialect) throws AdapterException {
-        PreparedStatement ps = null;
+            final String pushdownQuery, final SqlDialect dialect) {
         final ExaConnectionInformation connectionInformation = JdbcAdapterProperties
                 .getConnectionInformation(meta.getProperties(), exaMeta);
-
-        Connection connection = null;
         try {
-            connection = establishConnection(connectionInformation);
+            Connection connection = establishConnection(connectionInformation);
```
Comment (redcatbear) [inline jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/JdbcAdapter.java]: Add `final`.

File: jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/JdbcAdapter.java [hunk 5]
```diff
@@ -342,22 +347,21 @@ private static String getUserAndPasswordForImport(ExaMetadata exaMeta, SchemaMet
...
-                ps.execute();
+            ResultSetMetaData metadata = null;
+            try (PreparedStatement ps = connection.prepareStatement(pushdownQuery)) {
                 metadata = ps.getMetaData();
                 if (metadata == null) {
-                    throw new SQLException("getMetaData() failed");
+                    ps.execute();
+                    metadata = ps.getMetaData();
+                    if (metadata == null) {
+                        throw new SQLException("getMetaData() failed");
```
Comment (redcatbear) [inline jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/JdbcAdapter.java]: "Unable to read source metadata trying to create description for source columns."

File: jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/KerberosUtils.java [hunk 6]
```diff
@@ -33,37 +33,28 @@ public static void configKerberos(String user, String pass) throws Exception {
...
+        String keytabPath = writePath(krbDir, confKeytab[1], "kt_", ".keytab");
         String jaasConfigPath = writeJaasConfig(krbDir, user, keytabPath);
         System.setProperty("java.security.auth.login.config", jaasConfigPath);
         System.setProperty("java.security.krb5.conf", krbConfPath);
         System.setProperty("javax.security.auth.useSubjectCredsOnly", "false");
     }
 
-    private static String writeKrbConf(File krbDir, String base64Conf) throws Exception {
-        File file = File.createTempFile("krb_", ".conf", krbDir);
+    private static String writePath(File krbDir, String confKeyTab, String prefix, String suffix) throws Exception {
```
Comment (redcatbear) [inline jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/KerberosUtils.java]: Add `final` to all parameters and method variables.

File: jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/KerberosUtils.java [hunk 7]
```diff
@@ -33,37 +33,28 @@ public static void configKerberos(String user, String pass) throws Exception {
...
+        try (FileOutputStream os = new FileOutputStream(file);) {
+            os.write(DatatypeConverter.parseBase64Binary(confKeyTab));
+        }
         return file.getCanonicalPath();
     }
 
     private static String writeJaasConfig(File krbDir, String princ, String keytabPath) throws Exception {
         File file = File.createTempFile("jaas_", ".conf", krbDir);
         file.deleteOnExit();
         String jaasData;
```
Comment (redcatbear) [inline jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/KerberosUtils.java]: Use `StringBuilder`;

File: jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/dialects/SqlDialects.java (hunk 2 above)
Comment (AnastasiiaSergienko) [reply jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/dialects/SqlDialects.java]: done

File: jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/dialects/SqlDialects.java (hunk 3 above)
Comment (AnastasiiaSergienko) [reply jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/dialects/SqlDialects.java]: done

File: jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/JdbcAdapter.java (hunk 4 above)
Comment (AnastasiiaSergienko) [reply jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/JdbcAdapter.java]: done

File: jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/KerberosUtils.java (hunk 6 above)
Comment (AnastasiiaSergienko) [reply jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/KerberosUtils.java]: done

File: jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/KerberosUtils.java (hunk 7 above)
Comment (AnastasiiaSergienko) [reply jdbc-adapter/virtualschema-jdbc-adapter/src/main/java/com/exasol/adapter/jdbc/KerberosUtils.java]: Can't do it safely without an unit test. Should be done in a separated branch.

**Pull request merged**